### PR TITLE
[algo] refactor: diffusion loss registration with worker-side adapters

### PR DESCRIPTION
**Pull request by wtomin**
## Summary

Refactors diffusion loss registration from **plain functions with manual batch wiring** in `losses.py` to **class-based, worker-side adapters** in `diffusion_algos.py`.
- `@register_diffusion_loss(name)` now registers `DiffusionLossFn` subclasses (instantiated at registration) instead of callables.
- Each adapter declares `required_model_output_keys` / `required_data_keys`, validates inputs via `validate_inputs()`, and returns `DiffusionLossResult`.
- Pure loss math (`compute_diffusion_loss_flow_grpo`, etc.) is unchanged; worker adapters wrap it.
- KL regularization moves into the registry as `"kl"` (`KLLossFunc`); `kl_penalty_image` is renamed to `kl_penalty`.
- `losses.py` no longer special-cases `grpo_guard` or inlines KL logic.
Built-in adapters: `FlowGRPOLossFunc`, `GRPOGuardLossFunc`, `KLLossFunc`.

### 1. Registry API (`verl_omni/trainer/diffusion/diffusion_algos.py`)

| Aspect | main | this branch |
|--------|------|-------------|
| Registered object | `Callable` function | `DiffusionLossFn` subclass (`cls()` stored in registry) |
| Call signature | `fn(old_log_prob, log_prob, advantages, config=...)` | `fn(config=..., model_output=..., data=...) -> DiffusionLossResult` |
| Return value | `(Tensor, dict)` | `DiffusionLossResult(loss, metrics, add_loss_metric=True)` |
| Input validation | none | `validate_inputs()` with readable `KeyError` |

Three built-in adapters are added:

- **`FlowGRPOLossFunc`** (`"flow_grpo"`) — wraps `compute_diffusion_loss_flow_grpo`
- **`GRPOGuardLossFunc`** (`"grpo_guard"`) — wraps `compute_diffusion_loss_grpo_guard`; extracts SDE-related tensors itself
- **`KLLossFunc`** (`"kl"`) — wraps `kl_penalty` (renamed from `kl_penalty_image`)

Pure math loss functions (`compute_diffusion_loss_flow_grpo`, etc.) are kept for unit tests and documentation.

### 2. Worker simplification (`verl_omni/workers/utils/losses.py`)

- Removes the `loss_mode == "grpo_guard"` branch and manual kwargs assembly.
- Unified flow: `get_diffusion_loss_fn(mode)` → `validate_inputs()` → `__call__()` → merge metrics.
- KL path uses `get_diffusion_loss_fn("kl")` instead of directly importing `kl_penalty_image`.
- `DiffusionLossResult.add_loss_metric` controls whether a top-level actor loss metric is emitted.

### 3. Metric naming (behavior change — check dashboards)

| main | this branch |
|------|-------------|
| `metrics["actor/pg_loss"]` | `metrics["actor/loss"]` (policy loss) |
| KL only via `kl_loss` / `kl_coef` | same + optional `metrics["actor/kl_loss"]` (weighted KL term) |

Internal algorithm metrics (`actor/ppo_kl`, `actor/pg_clipfrac`, etc.) are unchanged.

### 4. Docs and tests

- `docs/contributing/integrating_a_new_algorithm_for_diffusion_model.md` — Step 3 updated: new algorithms must implement both a pure loss function and a `@register_diffusion_loss` adapter class.
- `docs/api/trainer.rst`, `docs/algo/grpo_guard.md` — symbol references updated.
- `tests/trainer/diffusion/test_diffusion_registry_on_cpu.py` — extended with:
  - built-in `"kl"` registration
  - `validate_inputs()` missing-key errors
  - custom loss registration via class pattern
  - `TestKLLossFunc` for the KL adapter

## Behavior change
- `metrics["actor/pg_loss"]` → `metrics["actor/loss"]`
- Optional `metrics["actor/kl_loss"]` when KL is enabled
## Test plan
- [x] `pytest tests/trainer/diffusion/test_diffusion_registry_on_cpu.py -v`
- [ ] Smoke: `loss_mode=flow_grpo`
- [ ] Smoke: `loss_mode=grpo_guard`
- [ ] Smoke: `use_kl_loss=True`

## Contributor integration pattern

```python
# 1. Pure loss (unit-testable)
def compute_diffusion_loss_my_algo(...) -> tuple[torch.Tensor, dict[str, Any]]:
    ...

# 2. Worker adapter (registered)
@register_diffusion_loss("my_algo")
class MyAlgoLossFunc(DiffusionLossFn):
    required_model_output_keys = ("log_probs",)
    required_data_keys = ("old_log_probs", "advantages")

    def __call__(self, *, config, model_output, data) -> DiffusionLossResult:
        loss, metrics = compute_diffusion_loss_my_algo(...)
        return DiffusionLossResult(loss=loss, metrics=metrics)

### CODE REVIEW

File: verl_omni/trainer/diffusion/diffusion_algos.py [hunk 1]
```diff
@@ -14,55 +14,91 @@
...
-    """
 
-    def decorator(func: DiffusionLossFn) -> DiffusionLossFn:
-        DIFFUSION_LOSS_REGISTRY[name] = func
-        return func
+class DiffusionLossFn:
+    """Base class for worker-side diffusion loss functions."""
 
-    return decorator
+    required_model_output_keys: tuple[str, ...] = ()
```
Comment (zhtmike) [inline verl_omni/trainer/diffusion/diffusion_algos.py:54]: needs to add comment what are `required_model_output_keys` and `required_data_keys`, why it is necessary.

File: verl_omni/pipelines/schedulers/flow_match_sde.py [hunk 2]
```diff
@@ -184,7 +184,7 @@ def sample_previous_step(
             return_sqrt_dt (`bool`, *optional*, defaults to False):
                 Whether to additionally return `sqrt(-dt)` as a tensor of shape `(batch_size,)`.
                 Used by GRPO-Guard to compute the importance-ratio normalization
-                (see `compute_diffusion_loss_grpo_guard`).
+                (see `GRPOGuardLossFunc`).
```
Comment (zhtmike) [inline verl_omni/pipelines/schedulers/flow_match_sde.py]: why don't name it as `GRPOGuardLoss` directly?

File: verl_omni/trainer/diffusion/diffusion_algos.py [hunk 3]
```diff
@@ -14,55 +14,91 @@
...
+    return "[" + ", ".join(keys) + "]"
 
-    Returns:
-        function: Decorator function that registers the diffusion loss function.
-    """
 
-    def decorator(func: DiffusionLossFn) -> DiffusionLossFn:
-        DIFFUSION_LOSS_REGISTRY[name] = func
-        return func
+class DiffusionLossFn:
```
Comment (zhtmike) [inline verl_omni/trainer/diffusion/diffusion_algos.py]: make it to be a `ABC`, will looks clearer

File: verl_omni/trainer/diffusion/diffusion_algos.py [hunk 4]
```diff
@@ -14,55 +14,91 @@
...
+        *,
+        config: DiffusionActorConfig,
+        model_output: dict[str, Any],
+        data: TensorDict,
+    ) -> DiffusionLossResult:
+        raise NotImplementedError
+
+
+DIFFUSION_LOSS_REGISTRY: dict[str, DiffusionLossFn] = {}
+DiffusionLossFnT = TypeVar("DiffusionLossFnT", bound=DiffusionLossFn)
```
Comment (zhtmike) [inline verl_omni/trainer/diffusion/diffusion_algos.py]: why uses `TypeVar`? I don't get it. ABC is clear enough.

File: verl_omni/trainer/diffusion/diffusion_algos.py [hunk 5]
```diff
@@ -247,7 +282,29 @@ def compute_diffusion_loss_flow_grpo(
     return pg_loss, pg_metrics
 
 
-@register_diffusion_loss("grpo_guard")
+@register_diffusion_loss("flow_grpo")
+class FlowGRPOLossFunc(DiffusionLossFn):
```
Comment (zhtmike) [inline verl_omni/trainer/diffusion/diffusion_algos.py]: 1. `FlowGRPOLossFunc` -> `FlowGRPOLoss`
2. make `compute_diffusion_loss_flow_grpo` inline. There suppose to be single loss function/class for single loss.
3. Same as `GRPOGuardLoss`

File: verl_omni/trainer/diffusion/diffusion_algos.py [hunk 6]
```diff
@@ -352,3 +436,32 @@ def kl_penalty_image(
     """
     kl_loss = ((prev_sample_mean - ref_prev_sample_mean) ** 2).mean(dim=(1, 2), keepdim=True) / (2 * std_dev_t**2)
     return kl_loss.mean()
+
+
+@register_diffusion_loss("kl")
+class KLLossFunc(DiffusionLossFn):
```
Comment (zhtmike) [inline verl_omni/trainer/diffusion/diffusion_algos.py]: `KLLoss` and make `kl_loss` inline

File: verl_omni/trainer/diffusion/diffusion_algos.py (hunk 1 above)
Comment (wtomin) [reply verl_omni/trainer/diffusion/diffusion_algos.py:54]: Docs added.

File: verl_omni/pipelines/schedulers/flow_match_sde.py (hunk 2 above)
Comment (wtomin) [reply verl_omni/pipelines/schedulers/flow_match_sde.py]: Rename is done.

File: verl_omni/trainer/diffusion/diffusion_algos.py (hunk 3 above)
Comment (wtomin) [reply verl_omni/trainer/diffusion/diffusion_algos.py]: Fixed.

**Review by AndyZhou952 — COMMENTED** [review]
One side remark: DiffusionNFT actually uses a different formulation of `kl_penalty`, my previous idea is to similarly use a `KL_LOSS_REGISTRY` and write a separate KL loss (maybe `kl_penalty_nft`). I wonder if this fits with our current `KLLoss` class?

File: verl_omni/trainer/diffusion/diffusion_algos.py (hunk 4 above)
Comment (wtomin) [reply verl_omni/trainer/diffusion/diffusion_algos.py]: Redundant. Removed.

File: verl_omni/trainer/diffusion/diffusion_algos.py (hunk 6 above)
Comment (wtomin) [reply verl_omni/trainer/diffusion/diffusion_algos.py]: Done

**Comment by wtomin** [comment]
> One side remark: DiffusionNFT actually uses a different formulation of `kl_penalty`, my previous idea is to similarly use a `KL_LOSS_REGISTRY` and write a separate KL loss (maybe `kl_penalty_nft`). I wonder if this fits with our current `KLLoss` class?

For simplicity, I will recommend you to rewrite exisiting `KLLoss(DiffusionLossFn)` to accommodate diffusion-nft. If it's very different, I think you can register a new kl loss with a different name.

File: verl_omni/trainer/diffusion/diffusion_algos.py (hunk 5 above)
Comment (wtomin) [reply verl_omni/trainer/diffusion/diffusion_algos.py]: Fixed.

File: verl_omni/trainer/diffusion/diffusion_algos.py [hunk 7]
```diff
@@ -13,56 +13,108 @@
...
+
+    @abstractmethod
+    def __call__(
+        self,
+        *,
+        config: DiffusionActorConfig,
+        model_output: dict[str, Any],
+        data: TensorDict,
+    ) -> DiffusionLossResult:
+        """Compute loss and metrics from the worker batch."""
```
Comment (zhtmike) [inline verl_omni/trainer/diffusion/diffusion_algos.py:101]: add a  `pass` ?

Comment (wtomin) [reply verl_omni/trainer/diffusion/diffusion_algos.py:101]: Add a `raise NotImplementedError`

**Review by zhtmike — APPROVED** [review]
